### ISSUE TITLE
Matching question answers can't be translated

### ISSUE THREAD

**Issue by jgramp** [andrewhancox/moodle-filter_translations#80]
I believe the old content translation plugin enabled us to translate answers in matching questions, by outputting the extra text in the drop down menus to the bottom of the page. Is there anything we can do to make the new plugin work with these types of questions? We use matching questions on our MEC site, so it'll be an issue if we can't in the new plugin.

![image](https://user-images.githubusercontent.com/8575694/164365604-4eb69e4b-caa4-4f24-b52f-cea8732f581f.png)

The languages menu allows language names to be translated, so perhaps there's a way for matching questions to be translated similarly:
![image](https://user-images.githubusercontent.com/8575694/164365480-b8caaf54-6f90-47eb-a9a5-708d7707564b.png)

**Comment by andrewhancox** [comment]
The two menus are constructed very differently - the languages menu is a bunch of markup laid out to look like a drop down, the options on the question is a select list which can only ever contain plain text. Assuming the items in this menu are filtered (they must be if this worked before) then you can pick this up using the 'missing on this page' report.

**Comment by jgramp** [comment]
It seems that the answers to matching questions are not run through a filter, and hence aren't translated:
I've tested and it's confirmed here: https://moodle.org/mod/forum/discuss.php?d=20759

We have many matching questions in the MEC Moodle that will need to be translated by the translation plugin.

The old plugin would output the plain text fields on the page below the content. How can we resolve this issue, so we can translate these matching questions? 
@rjnl any ideas? @andrewhancox - can you suggest a way we might implement this please?

**Comment by andrewhancox** [comment]
This system is not able to do anything about un-filtered text and the previous filter would not have been able to either.
It’s probably not what you want to hear but the fix here is to get the question type to correctly implement filters.

**Comment by annakrassa** [comment]
Hi everyone, 
The old plugin allowed us to translate the options in a matching question, not through the question, but through the list of items displayed in the page. 
![Screenshot (3)](https://user-images.githubusercontent.com/100235318/168137738-8d7ebb37-de43-4590-a3f6-37bc86357b1d.png)

**Comment by andrewhancox** [comment]
If the question options are run through the filter system properly then you will be able to translate them by going to the missed translations list - if not then additional work will need to be done to find a way to handle this use case.

**Comment by jgramp** [comment]
Yes, I've confirmed this now works, now we have a search feature to find
them in the missing list. So Anna, rest assured, there is a way
to translate the matching questions, but it's not pretty, so Rajneel will
look at making this easier for our editors next week.
I have proposed a solution to Rajneel which he's not seen yet.

I proposed:
Would there maybe be a way to check if the page URL is a quiz attempt and
then if it is, search the translations database for that page (URL) and
search string for each drop down menu item (original content??) and then
recreate the translation edit buttons (stale, missing, ok) that way? Would
these go in the drop down menu, like they do for the language menu in the
Moodle header?

Jessica Gramp
Moodle Academy Manager
UTC+10:30 (until 3 Apr 2022)


***@***.***
moodle.com
[image: Twitter] <https://twitter.com/moodle> [image: Facebook]
<https://www.facebook.com/moodle/> [image: Youtube]
<https://www.youtube.com/moodlehq> [image: Moodle - Empowering educators to
improve our world] <https://moodle.com/>
[image: Moodle Academy: Develop your skills as a Moodle Edcuator,
Administrator, Designer and Developer.] <https://moodle.academy>

**Comment by andrewhancox** [comment]
I would not start adding rules to the filter that apply to specific plugins or the plugin will become unmanageable.

The reason the translation buttons are not displayed is that the filtered text is being shown as an option in an HTML select element so cannot contain markup - the language menu is a bootstrap drop down menu so can have rich content in it.
My suspicion is that there may be a way to detect if any translation buttons have been stripped and display them in some other way on the page... It's likely that there will be a few other situations that will come to light where this is happening.

**Comment by jgramp** [comment]
Ok great .Thanks Andrew, I'll let Rajneel come up with a solution in that
case. If you think of any ways this might work, please let him know.
Jessica Gramp
Moodle Academy Manager
UTC+10:30 (until 3 Apr 2022)


***@***.***
moodle.com
[image: Twitter] <https://twitter.com/moodle> [image: Facebook]
<https://www.facebook.com/moodle/> [image: Youtube]
<https://www.youtube.com/moodlehq> [image: Moodle - Empowering educators to
improve our world] <https://moodle.com/>
[image: Moodle Academy: Develop your skills as a Moodle Edcuator,
Administrator, Designer and Developer.] <https://moodle.academy>

**Comment by berthelemy** [comment]
@jgramp We're encountering a similar issue with drag and drop matching questions. I'm guessing it's the same problem - that the answers aren't run through a filter?

Did you come up with a solution?
